File: pointcloud_transformer_cpp/src/gng_transformer_node.cpp
#include <algorithm>
#include <cmath>
#include <memory>
#include <string>
#include <unordered_map>
#include <vector>

#include "ais_gng_msgs/msg/topological_map.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2_eigen/tf2_eigen.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"
#include <Eigen/Geometry>

/**
 * @brief High-performance GNG Transformer Node.
 * Transforms a TopologicalMap from sensor frame to target frame (e.g.,
 * base_link) using Eigen matrix operations for maximum speed. Supports optional
 * workspace filtering.
 */
class GngTransformerNode : public rclcpp::Node {
public:
  GngTransformerNode() : Node("gng_transformer_node") {
    // Declare and get parameters
    this->declare_parameter<std::string>("target_frame", "base_link");
    this->declare_parameter<std::string>("input_topic", "/gng_map");
    this->declare_parameter<std::string>("output_topic",
                                         "/topological_map_transformed");
    this->declare_parameter<double>("filter_radius",
                                    -1.0); // Negative to disable
    this->declare_parameter<double>("filter_z_min", -1.0);
    this->declare_parameter<double>("filter_z_max", 1.0);

    target_frame_ = this->get_parameter("target_frame").as_string();
    filter_radius_ = this->get_parameter("filter_radius").as_double();
    filter_z_min_ = this->get_parameter("filter_z_min").as_double();
    filter_z_max_ = this->get_parameter("filter_z_max").as_double();

    auto input_topic = this->get_parameter("input_topic").as_string();
    auto output_topic = this->get_parameter("output_topic").as_string();

    // TF2 listener and buffer
    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

    // Subscription
    subscription_ =
        this->create_subscription<ais_gng_msgs::msg::TopologicalMap>(
            input_topic, 10,
            std::bind(&GngTransformerNode::map_callback, this,
                      std::placeholders::_1));

    // Publisher
    publisher_ = this->create_publisher<ais_gng_msgs::msg::TopologicalMap>(
        output_topic, 10);

    RCLCPP_INFO(this->get_logger(),
                "GNG Transformer started: %s -> %s (Target: %s, Filter: %.2f)",
                input_topic.c_str(), output_topic.c_str(),
                target_frame_.c_str(), filter_radius_);
  }

private:
  void map_callback(const ais_gng_msgs::msg::TopologicalMap::SharedPtr msg) {
    if (msg->nodes.empty())
      return;

    geometry_msgs::msg::TransformStamped tf_msg;
    try {
      tf_msg = tf_buffer_->lookupTransform(target_frame_, msg->header.frame_id,
                                           msg->header.stamp,
                                           rclcpp::Duration::from_seconds(0.1));
    } catch (const tf2::TransformException &ex) {
      RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
                           "Could not transform %s to %s: %s",
                           msg->header.frame_id.c_str(), target_frame_.c_str(),
                           ex.what());
      return;
    }

    Eigen::Isometry3d transform = tf2::transformToEigen(tf_msg);
    Eigen::Matrix3d rotation = transform.rotation();

    auto transformed_map =
        std::make_unique<ais_gng_msgs::msg::TopologicalMap>();
    transformed_map->header.stamp = msg->header.stamp;
    transformed_map->header.frame_id = target_frame_;

    std::unordered_map<uint16_t, uint16_t> old_to_new_id;
    const double r2 = filter_radius_ * filter_radius_;

    // 1. Transform and Filter Nodes
    for (const auto &node : msg->nodes) {
      Eigen::Vector3d pos_old(node.pos.x, node.pos.y, node.pos.z);
      Eigen::Vector3d pos_new = transform * pos_old;

      // Optional workspace filtering
      if (filter_radius_ > 0.0) {
        double dist2 = pos_new.x() * pos_new.x() + pos_new.y() * pos_new.y();
        if (dist2 > r2 || pos_new.z() < filter_z_min_ ||
            pos_new.z() > filter_z_max_) {
          continue;
        }
      }

      ais_gng_msgs::msg::TopologicalNode new_node = node;
      new_node.pos.x = pos_new.x();
      new_node.pos.y = pos_new.y();
      new_node.pos.z = pos_new.z();

      // Rotate normal vector
      Eigen::Vector3d normal_old(node.normal.x, node.normal.y, node.normal.z);
      Eigen::Vector3d normal_new = rotation * normal_old;
      new_node.normal.x = normal_new.x();
      new_node.normal.y = normal_new.y();
      new_node.normal.z = normal_new.z();

      old_to_new_id[node.id] = transformed_map->nodes.size();
      transformed_map->nodes.push_back(new_node);
    }

    // 2. Filter Edges
    // Edges are pairs of node indices.
    for (size_t i = 0; i + 1 < msg->edges.size(); i += 2) {
      uint16_t id1 = msg->edges[i];
      uint16_t id2 = msg->edges[i + 1];

      if (old_to_new_id.count(id1) && old_to_new_id.count(id2)) {
        transformed_map->edges.push_back(old_to_new_id[id1]);
        transformed_map->edges.push_back(old_to_new_id[id2]);
      }
    }

    // 3. Keep clusters (optional, depends on implementation)
    transformed_map->clusters = msg->clusters;

    publisher_->publish(std::move(transformed_map));
  }

  std::string target_frame_;
  double filter_radius_;
  double filter_z_min_;
  double filter_z_max_;

  std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
  std::shared_ptr<tf2_ros::TransformListener> tf_listener_{nullptr};
  rclcpp::Subscription<ais_gng_msgs::msg::TopologicalMap>::SharedPtr
      subscription_;
  rclcpp::Publisher<ais_gng_msgs::msg::TopologicalMap>::SharedPtr publisher_;
};

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<GngTransformerNode>());
  rclcpp::shutdown();
  return 0;
}

File: pointcloud_transformer/pointcloud_transformer/gng_transformer_node.py
import rclpy
from rclpy.node import Node
import tf2_ros
from tf2_ros import TransformException
import tf2_geometry_msgs  # This is key for transforming geometry_msgs types

# Assuming the message definitions are from a package named 'ais_gng_msgs'
# You might need to adjust this based on your actual package structure
from ais_gng_msgs.msg import TopologicalMap, TopologicalNode
from geometry_msgs.msg import PointStamped, Point32


class GngTransformerNode(Node):
    def __init__(self):
        super().__init__("gng_transformer_node")

        # Declare parameters
        self.declare_parameter("target_frame", "arm_base_link")
        self.declare_parameter("input_topic", "/gng_map")
        self.declare_parameter("output_topic", "/gng_map_transformed")

        # Get parameters
        self.target_frame = (
            self.get_parameter("target_frame").get_parameter_value().string_value
        )
        input_topic = (
            self.get_parameter("input_topic").get_parameter_value().string_value
        )
        output_topic = (
            self.get_parameter("output_topic").get_parameter_value().string_value
        )

        # TF2 listener and buffer
        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self)

        # Subscription
        self.subscription = self.create_subscription(
            TopologicalMap, input_topic, self.map_callback, 10
        )

        # Publisher
        self.publisher = self.create_publisher(TopologicalMap, output_topic, 10)

        self.get_logger().info(
            f"GNG Transformer Node started. "
            f"Listening on '{input_topic}', "
            f"publishing on '{output_topic}', "
            f"transforming to '{self.target_frame}'."
        )

    def map_callback(self, msg: TopologicalMap):
        try:
            # 1. Look up the transform from the source frame to the target frame
            transform = self.tf_buffer.lookup_transform(
                self.target_frame,
                msg.header.frame_id,
                msg.header.stamp,
                timeout=rclpy.duration.Duration(seconds=0.1),
            )

            # 2. Create a new map to hold the transformed data
            transformed_map = TopologicalMap()
            transformed_map.header.stamp = msg.header.stamp
            transformed_map.header.frame_id = self.target_frame  # Update the frame_id!
            transformed_map.edges = msg.edges
            transformed_map.clusters = msg.clusters

            # 3. Transform each node's position
            for node in msg.nodes:
                # To use do_transform_point, we need a PointStamped
                point_stamped = PointStamped()
                point_stamped.header = msg.header
                # Point32 doesn't have 'z', so handle potential conversion issues if needed
                point_stamped.point.x = float(node.pos.x)
                point_stamped.point.y = float(node.pos.y)
                point_stamped.point.z = float(node.pos.z)

                # Perform the transformation
                transformed_point_stamped = tf2_geometry_msgs.do_transform_point(
                    point_stamped, transform
                )

                # Create a new node with the transformed position
                new_node = TopologicalNode()
                new_node.id = node.id
                new_node.pos = Point32(
                    x=transformed_point_stamped.point.x,
                    y=transformed_point_stamped.point.y,
                    z=transformed_point_stamped.point.z,
                )
                new_node.normal = node.normal
                new_node.rho = node.rho
                new_node.label = node.label
                new_node.age = node.age
                new_node.inpcl_ids = node.inpcl_ids

                transformed_map.nodes.append(new_node)

            # 4. Publish the transformed map
            self.publisher.publish(transformed_map)

        except TransformException as ex:
            self.get_logger().warn(
                f"Could not transform {msg.header.frame_id} to {self.target_frame}: {ex}",
                throttle_duration_sec=1.0,  # Avoid spamming logs
            )
        except Exception as e:
            self.get_logger().error(f"An unexpected error occurred: {e}")


def main(args=None):
    rclpy.init(args=args)
    node = GngTransformerNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
